common/s3util.cpp

```diff
@@ -41,11 +50,77 @@ using Aws::S3::Model::ListObjectsRequest;
 using Aws::S3::Model::HeadObjectRequest;
 using Aws::Utils::RateLimits::DefaultRateLimiter;
 
+
 namespace common {
 
+//TODO(angxu) currently page size is used as buffer size,
+// we might want to use a larger buffer size if write is
+// too slow
+const uint32_t kPageSize = getpagesize();
+
+DirectIOFile::DirectIOFile(const string& file_path)
+    : fd_(-1)
+    , file_size_(0)
+    , buffer_()
+    , offset_(0) {
+  if (posix_memalign(&buffer_, kPageSize, kPageSize) != 0) {
+    LOG(ERROR) << "Failed to allocate memaligned buffer, errno = " << errno;
+    return;
+  }
+
+  int flag = O_WRONLY | O_TRUNC | O_CREAT | O_DIRECT;
+  fd_ = open(file_path.c_str(), flag , 640);
+  if (fd_ < 0) {
+    LOG(ERROR) << "Failed to open " << file_path << " with flag " << flag
+               << ", errno = " << errno;
+    free(buffer_);
+    buffer_ = nullptr;
+  }
+}
+
+DirectIOFile::~DirectIOFile() {
+  if (buffer_ == nullptr || fd_ < 0) {
+    return;
+  }
+
+  if (offset_ > 0) {
+    if (::write(fd_, buffer_, kPageSize) != kPageSize) {
+      LOG(ERROR) << "Failed to write last chunk, errno = " << errno;
+    } else {
+      ftruncate(fd_, file_size_);
+    }
+  }
+  free(buffer_);
+  close(fd_);
+}
+
+std::streamsize DirectIOFile::write(const char* s, std::streamsize n) {
+  if (buffer_ == nullptr || fd_ < 0) {
+    return -1;
+  }
+  uint32_t remaining = static_cast<uint32_t>(n);
+  while (remaining > 0) {
+    uint32_t bytes = std::min((kPageSize - offset_), remaining);
+    std::memcpy((char*)buffer_ + offset_, s, bytes);
```

Comment:
It seems hard to catch this in test because `boost::iostreams::stream` is probably doing some internal buffering and never write buffer larger 4096...